src/HttpCache/TagCollector.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\HttpCache;
+
+use ApiPlatform\Metadata\ApiProperty;
+use Symfony\Component\PropertyInfo\Type;
+
+/**
+ * Collects cache tags during normalization.
+ *
+ * @author Urban Suppiger <urban@suppiger.net>
+ */
+class TagCollector implements TagCollectorInterface
```

Comment:
can you explain to me the difference between all these methods? I think that we should be able to simplify this and reduce the number of arguments. I'd suggest:

```
use ApiPlatform\Serializer; // we want to avoid creating a dependency between Serializer and HttpCache, the opposite is fine.

interface IriCollectorInterface {
    public function collect(string $iri, mixed $object, array $context) {}
}
```

Not even sure if `object` matters in the signature and it could be added to the context? It'd be interesting to add the `attribute` that's being normalized if needed although that I'd start without it and if someone needs that we add it afterwards? Or do you already have a use case where that matters? 

If that's the case, we can put everything we need inside our context. I'd not have a `childContext` though, only give the current context it introduces less complexity and the use cases are often not good ones.

Reply:
The various methods (extension points) are needed in order to implement the cache logic proposed under https://github.com/api-platform/core/pull/5650 in userland.

In order to implement this logic, the tag collector service on user side would look something like below (no guarantee for correctness, typing in a text editor only).

```php
<?php

class CustomTagCollector implements TagCollectorInterface
{
    public function collectTagsFromNormalize(mixed $object, string $format = null, array $context = [], string $iri = null): void
    {
        $this->addCacheTagForRessource($context, $iri);
    }

    public function collectTagsFromNormalizeRelation(mixed $object, string $format = null, array $context = [], string $iri = null): void
    {
    }

    public function collectTagsFromGetAttribute(mixed $object, string $format = null, array $context = [], string $iri = null, string $attribute = null, ApiProperty $propertyMetadata = null, Type $type = null, array $childContext = []): void
    {
        $this->addCacheTagsForRelation(array $context, $propertyMetadata)
    }

    private function addCacheTagForRessource(array $context, ?string $iri): void
    {
        if (isset($context['resources']) && isset($iri)) {
            $context['resources'][$iri] = $iri;
        }
    }
    
    private function addCacheTagsForRelation(array $context, ApiProperty $propertyMetadata): void
    {
        // Add cache tag for related collection
        if (isset($context['resources'])) {
            if (isset($propertyMetadata->getExtraProperties()['cacheDependencies'])) {
                foreach ($propertyMetadata->getExtraProperties()['cacheDependencies'] as $dependency) {
                    $cacheTag = $context['iri'].self::IRI_RELATION_DELIMITER.$dependency;
                    $context['resources'][$cacheTag] = $cacheTag;
                }
            } else {
                $cacheTag = $context['iri'].self::IRI_RELATION_DELIMITER.$context['api_attribute'];
                $context['resources'][$cacheTag] = $cacheTag;
            }
        }
    }
    
}
```

If there's only one `collect` method, I'd need to figure out from which point in the normalization process the collect method was called. Might be possible based on the content of the attributes, but not sure whether this leads to robust and understandable code.

With regards to parameters, in order to implement the original behavior of api-platform as well as the logic from https://github.com/api-platform/core/pull/5650 I'd need the following parameters as a minimum: `$context`, `$iri`, `$propertyMetadata`

All the rest (`$object`, `$format`, `$childContext`, `$attribute`, `$type`) can also be removed and we could add in case there's a specific usecase coming up for it. `$type` would be needed though, if anyone wants to differentiate relations to collections (*ToMany) and *ToOne relations.

Reply:
No prob, fully understand design of the interface is crucial.

For implementation of the current api-platform logic, I could use `$attribute` to ignore calls coming from `getAttribute`, because this parameter is not sent from `normalize` or `normalizeRelation`. I cannot use `$context['api_attribute']`, because this  is set for nested/embedded objects also for calls from `normalize` or `normalizeRelation`.

```php
class TagCollector implements TagCollectorInterface
{
    public function collect(mixed $object, string $format = null, array $context = [], string $iri = null, string $attribute = null, ApiProperty $propertyMetadata = null, Type $type = null, array $childContext = []): void
    {
        if(!$attribute){
            $this->addResourceToContext($context, $iri);
        }
    }

    private function addResourceToContext(array $context, ?string $iri): void
    {
        if (isset($context['resources']) && isset($iri)) {
            $context['resources'][$iri] = $iri;
        }
    }
}
```

Implementing the logic above for #5650 is a bit more tricky. With the current code, I don't see any good means to differentiate between calls from `normalize` and `normalizeRelation`. I could add an additional string attribute to disclose, where the collect call was made from. But honestly, that seems a bit like a disguised event system and I'd rather have a proper event system implemented as in the initial PR #5753.

Alternatively I think I could add the normalized `$data` back as an attribute and then check within collect, if data is an object (fully normalized embedded resource) or only a string (resource normalized as IRI only).

Reply:
You need to find out whether the call is made in a `normalizeRelation` as opposed to a `normalize`? you have `api_sub_level` for that IIRC.

Reply:
> You need to find out whether the call is made in a `normalizeRelation` as opposed to a `normalize`? 

Yeah, correct. https://github.com/api-platform/core/pull/5650 gives more background on the motivation. But in a nutshell, I'd want to treat fully embedded resources (resource embedded in the response as object) differently from linked resources (only IRI of resource included in response).

For the former, the cache tag is triggered from `normalize`. For latter, the cache tag is triggered from `normalizeRelation` (at least that's how it's implemented now).

> you have `api_sub_level` for that IIRC.

Not sure, this works. When running tags.feature tests, `api_sub_level` seems always to be true.

Reply:
> Alternatively I think I could add the normalized `$data` back as an attribute and then check within collect, if data is an object (fully normalized embedded resource) or only a string (resource normalized as IRI only).

I just pushed a proposal implementing this. When passing `$data`, I can check on userland if `$data` is an array/object or a string (IRI) only and hence differentiate embedded objects from linked IRIs.

For JsonAPI this is a bit "trickier", because even for linked data, the [return value is an array](https://github.com/api-platform/core/pull/5758/files#diff-fe1eba7bf34edf60f0eefbfb87fda8fb5aff8f676ab95af88c3ffe276db8a253R241) and not only a string. However, as we pass `$format` as well, this could still be differentiated from a normal fully embedded object.

I had to change the tests, because now the order of tags has changed when waiting for the normalized data. The tags are still the same though.

Reply:
Please resolve https://github.com/api-platform/core/pull/5758#discussion_r1307370466 so that we don't change any tests